File: mnistTF.js
import * as tf from '@tensorflow/tfjs-node-gpu';  // Use TensorFlow.js with GPU
import * as fs from 'fs';
import * as path from 'path';
import * as zlib from 'zlib';

// Check if GPU is available
if (tf.ENV.get('WEBGL_RENDER_FLOAT32_CAPABLE')) {
  console.log('TensorFlow.js is using WebGL for GPU acceleration.');
} else {
  console.log('TensorFlow.js is not using GPU. Press enter to continue.');
}

// Directory structure for MNIST data
const mnistDir = './data/MNIST/raw';

function loadMNISTImages(filename) {
  const buffer = zlib.gunzipSync(fs.readFileSync(filename));  // Decompress the .gz file
  const magic = buffer.readUInt32BE(0);
  const numImages = buffer.readUInt32BE(4);
  const rows = buffer.readUInt32BE(8);
  const cols = buffer.readUInt32BE(12);
  const imageData = buffer.slice(16);

  const images = new Float32Array(numImages * rows * cols);
  for (let i = 0; i < imageData.length; i++) {
    images[i] = imageData[i] / 255.0;  // Normalize to 0-1
  }

  return tf.tensor4d(images, [numImages, rows, cols, 1]);  // Shape (numImages, 28, 28, 1)
}
function loadMNISTLabels(filename) {
    const buffer = zlib.gunzipSync(fs.readFileSync(filename));  // Decompress the .gz file
    const labels = buffer.slice(8);  // Skip the magic number
    return tf.tensor1d(labels, 'int32').toFloat();  // Convert to float32 tensor
  }
  

// Load train and test data
const xTrain = loadMNISTImages(path.join(mnistDir, 'train-images-idx3-ubyte.gz'));
const yTrain = loadMNISTLabels(path.join(mnistDir, 'train-labels-idx1-ubyte.gz'));
const xTest = loadMNISTImages(path.join(mnistDir, 't10k-images-idx3-ubyte.gz'));
const yTest = loadMNISTLabels(path.join(mnistDir, 't10k-labels-idx1-ubyte.gz'));

// Build the model
const model = tf.sequential();
model.add(tf.layers.flatten({inputShape: [28, 28, 1]}));  // Flatten input
model.add(tf.layers.dense({units: 128, activation: 'relu'}));  // Fully connected layer
model.add(tf.layers.dropout({rate: 0.2}));  // Dropout layer
model.add(tf.layers.dense({units: 10, activation: 'softmax'}));  // Output layer

// Compile the model
model.compile({
  optimizer: 'adam',
  loss: 'sparseCategoricalCrossentropy',
  metrics: ['accuracy']
});

// Train the model
async function trainModel() {
  await model.fit(xTrain, yTrain, {
    epochs: 10,
    validationData: [xTest, yTest],
  });

  // Evaluate the model
  const evalResult = model.evaluate(xTest, yTest);
  const testAcc = evalResult[1].dataSync();
  console.log(`Test accuracy: ${testAcc}`);

  // Save the model to TensorFlow.js format
  const tfjsTargetDir = './model/web_model';
  await model.save(`file://${tfjsTargetDir}`);
  console.log(`Model saved to ${tfjsTargetDir}`);
}

// Start training
trainModel();

File: mnistTF.py
import tensorflow as tf
from tensorflow.keras import layers, models
import numpy as np
import os
import time
import gzip
import struct
import tensorflowjs as tfjs

startTime = time.time()

# Check if TensorFlow is using a GPU
if tf.config.list_physical_devices('GPU'):
    print("TensorFlow is using the following GPU(s):")
    for gpu in tf.config.list_physical_devices('GPU'):
        print(gpu)
else:
    print("TensorFlow is not using a GPU.")
    input("Press enter to continue")

# Directory structure for MNIST data
mnist_dir = './data/MNIST/raw'

def load_mnist_images(filename):
    with gzip.open(filename, 'rb') as f:
        _, num_images, rows, cols = struct.unpack(">IIII", f.read(16))
        data = np.frombuffer(f.read(), dtype=np.uint8).reshape(num_images, rows, cols)
        return data

def load_mnist_labels(filename):
    with gzip.open(filename, 'rb') as f:
        f.read(8)  # skip the magic number and number of labels
        labels = np.frombuffer(f.read(), dtype=np.uint8)
        return labels

# Load train and test data
x_train = load_mnist_images(os.path.join(mnist_dir, 'train-images-idx3-ubyte.gz'))
y_train = load_mnist_labels(os.path.join(mnist_dir, 'train-labels-idx1-ubyte.gz'))
x_test = load_mnist_images(os.path.join(mnist_dir, 't10k-images-idx3-ubyte.gz'))
y_test = load_mnist_labels(os.path.join(mnist_dir, 't10k-labels-idx1-ubyte.gz'))

# Normalize the pixel values to be between 0 and 1
x_train, x_test = x_train / 255.0, x_test / 255.0

# Build the model
model = models.Sequential([
    layers.Flatten(input_shape=(28, 28)),  # Flatten the 28x28 images into 1D vectors
    layers.Dense(128, activation='relu'),  # Fully connected layer with 128 units and ReLU activation
    layers.Dropout(0.2),  # Dropout layer to reduce overfitting
    layers.Dense(10, activation='softmax')  # Output layer with 10 units (one for each class) and softmax activation
])

# Compile the model
model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

# Train the model
model.fit(x_train, y_train, epochs=10)

# Evaluate the model
test_loss, test_acc = model.evaluate(x_test, y_test)

print(f"Test accuracy: {test_acc:.4f}")
print(f"Elapsed time: {time.time()-startTime}")

saveModel = True
if saveModel:
    # Save the model in TensorFlow.js format
    tfjs_target_dir = './model/web_model'
    tfjs.converters.save_keras_model(model, tfjs_target_dir)

    print(f"Model saved to {tfjs_target_dir}")
